[LoopFusion] Detecting legal dependencies for fusion using DA info

1997alireza opened the pull request:
Using the information provided by the recent DA patch to detect additional legal fusions.

Meinersbur commented:
Stacked PRs are applied to the branch of the predecessor PR (https://github.com/1997alireza/llvm-project/tree/da-siv-separate-loops). However, problems occur when branches are PRs are from different repositories (here: https://github.com/llvm/llvm-project and https://github.com/1997alireza/llvm-project).

Anyway, one can see the changes introduced only by this PR here:
https://github.com/1997alireza/llvm-project/compare/da-siv-separate-loops...loopfuse-da-info

Meinersbur commented on llvm/test/Transforms/LoopFusion/simple.ll:
```diff
@@ -300,40 +300,36 @@ bb23:                                             ; preds = %bb17, %bb
 ; CHECK-LABEL: @forward_dep(
-; CHECK-NEXT:  bb:
-; CHECK-NEXT:    br label [[BB7:%.*]]
+; CHECK-NEXT:  [[BB:.*]]:
```
Could you add some pseudo-C source for this test? Atm I cannot see whether it now being fused is ok.

1997alireza replied:
Added.

1997alireza committed:
[LoopFusion] Detecting legal dependencies for fusion using DA info

Loop fusion pass will use the information provided by the recent
DA patch to fuse additional legal loops, including those with
forward loop-carried dependencies.

amehsan commented on llvm/include/llvm/Analysis/DependenceAnalysis.h:
```diff
@@ -82,6 +82,16 @@ class LLVM_ABI Dependence {
   /// Dependence::DVEntry - Each level in the distance/direction vector
   /// has a direction (or perhaps a union of several directions), and
   /// perhaps a distance.
+  /// The dependency information could be across a single loop level or across
```
The changes in DA are already merged in a different PR. Is that correct? If yes, you need to rebase this.

amehsan commented on llvm/test/Transforms/LoopFusion/backward_loop_carried.ll:
```diff
@@ -0,0 +1,185 @@
+;;      for (long int k = 0; k < n; k++) {
+;;        A[i][j][k] = i;
+;;      }
+;;      for (long int k = 0; k < n; k++) {
+;;        temp = A[i + 3][j + 2][k + 1];
+
+define void @backward_dep0(i64 %n, ptr %A) nounwind uwtable ssp {
```
Why C code is incomplete?

amehsan commented on llvm/test/Transforms/LoopFusion/backward_loop_carried.ll:
```diff
@@ -0,0 +1,185 @@
+;;      for (long int k = 0; k < n; k++) {
+;;        A[i][j][k] = i;
+;;      }
+;;      for (long int k = 0; k < n; k++) {
+;;        temp = A[i][j][k - 1];
+
+define void @backward_dep1(i64 %n, ptr %A) nounwind uwtable ssp {
```
Similarly incomplete C code

1997alireza replied:
Fixed.

1997alireza commented on llvm/lib/Transforms/Scalar/LoopFuse.cpp:
```diff
@@ -1371,6 +1373,33 @@ struct LoopFuser {
+
+      assert(CurLoopLevel > Levels && "Fusion candidates are not separated");
+      unsigned CurDir = DepResult->getDirection(CurLoopLevel, true);
+      if (!(CurDir & Dependence::DVEntry::GT)) {
```
Similar to above.

CongzheUalberta commented on llvm/lib/Transforms/Scalar/LoopFuse.cpp:
```diff
@@ -100,6 +100,8 @@ STATISTIC(OnlySecondCandidateIsGuarded,
           "The second candidate is guarded while the first one is not");
 STATISTIC(NumHoistedInsts, "Number of hoisted preheader instructions.");
 STATISTIC(NumSunkInsts, "Number of hoisted preheader instructions.");
+STATISTIC(NumDepSafeFused, "Number of fused loops with dependencies "
```
I suggest to change the description to something more descriptive, something like "dependence checks passed".

CongzheUalberta commented on llvm/lib/Transforms/Scalar/LoopFuse.cpp:
```diff
@@ -1371,6 +1373,33 @@ struct LoopFuser {
+      unsigned SameSDLevels = DepResult->getSameSDLevels();
+      unsigned CurLoopLevel = FC0.L->getLoopDepth();
+
+      bool OuterEqDir = true;
```
I suggest to rename `OuterEqDir` to something like `sameOuterIndex` or `sameOuterIterSpace` that is more descriptive.

CongzheUalberta commented on llvm/test/Transforms/LoopFusion/backward_loop_carried.ll:
```diff
@@ -0,0 +1,187 @@
+;;    }
+;;  }
+
+define void @backward_dep0(i64 %n, ptr %A) nounwind uwtable ssp {
```
I suggest to rename this function. This function is testing the case where accesses to `A` in the outer dimensions are different, hence in terms of the innermost dimension `k`, there is no dependency on the `k-level`.

CongzheUalberta commented on llvm/test/Transforms/LoopFusion/backward_loop_carried.ll:
```diff
@@ -0,0 +1,187 @@
+  br i1 %cmp10, label %for.cond1.preheader.preheader, label %for.end26
+
+; CHECK-LABEL: backward_dep
+; CHECK-COUNT-1: for.body{{[0-9]+}}:
```
I suggest to not rely on BB names in `CHECK`. You can check `stats`, like how many functions are fused, how many dependence checks are passed, etc.

CongzheUalberta commented on llvm/test/Transforms/LoopFusion/backward_loop_carried.ll:
```diff
@@ -0,0 +1,187 @@
+  br i1 %cmp10, label %for.cond1.preheader.preheader, label %for.end26
+
+; CHECK-LABEL: backward_dep
+; CHECK-COUNT-1: for.body{{[0-9]+}}:
```
Same as the comment above.

CongzheUalberta commented on llvm/test/Transforms/LoopFusion/simple.ll:
```diff
@@ -298,42 +298,51 @@ bb23:                                             ; preds = %bb17, %bb
+;         arg[j] = val * 3;
+;     }
+; }
+;
```
This function was not fused previously, and now with your patch it is fused. Can you add some comment here to describe it like why it can be fused now?

1997alireza commented on llvm/lib/Transforms/Scalar/LoopFuse.cpp:
```diff
@@ -1371,6 +1373,33 @@ struct LoopFuser {
+      unsigned CurLoopLevel = FC0.L->getLoopDepth();
+
+      bool OuterEqDir = true;
+      for (unsigned II = 1; II <= std::min(CurLoopLevel - 1, Levels); ++II) {
```
Changed to Level.

CongzheUalberta reviewed (approved):
Thanks for addressing the comments. LGTM.
